src/objective/rank_obj.cu

```diff
@@ -465,105 +525,100 @@ struct NDCGLambdaWeightComputer {
+#endif
+{
+ public:
   struct MAPStats {
     /*! \brief the accumulated precision */
-    float ap_acc;
+    float ap_acc{0.0f};
     /*!
      * \brief the accumulated precision,
      *   assuming a positive instance is missing
      */
-    float ap_acc_miss;
+    float ap_acc_miss{0.0f};
     /*!
      * \brief the accumulated precision,
      * assuming that one more positive instance is inserted ahead
      */
-    float ap_acc_add;
+    float ap_acc_add{0.0f};
     /* \brief the accumulated positive instance count */
-    float hits;
-    MAPStats() = default;
-    MAPStats(float ap_acc, float ap_acc_miss, float ap_acc_add, float hits)
-        : ap_acc(ap_acc), ap_acc_miss(ap_acc_miss), ap_acc_add(ap_acc_add), hits(hits) {}
+    float hits{0.0f};
+
+    XGBOOST_DEVICE MAPStats() {}  // NOLINT
+    XGBOOST_DEVICE MAPStats(float ap_acc, float ap_acc_miss, float ap_acc_add, float hits)
+      : ap_acc(ap_acc), ap_acc_miss(ap_acc_miss), ap_acc_add(ap_acc_add), hits(hits) {}
+
+    // For prefix scan
+    XGBOOST_DEVICE MAPStats operator +(const MAPStats &v1) const {
+      return {ap_acc + v1.ap_acc, ap_acc_miss + v1.ap_acc_miss,
+              ap_acc_add + v1.ap_acc_add, hits + v1.hits};
+    }
+
+    // For test purposes - compare for equality
+    XGBOOST_DEVICE bool operator ==(const MAPStats &rhs) const {
+      return ap_acc == rhs.ap_acc && ap_acc_miss == rhs.ap_acc_miss &&
+             ap_acc_add == rhs.ap_acc_add && hits == rhs.hits;
+    }
   };
 
+ private:
+  template <typename T>
+  XGBOOST_DEVICE inline static void Swap(T &v0, T &v1) {
+#if defined(__CUDACC__)
+    thrust::swap(v0, v1);
+#else
+    std::swap(v0, v1);
+#endif
+  }
+
   /*!
-   * \brief Obtain the delta MAP if trying to switch the positions of instances in index1 or index2
-   *        in sorted triples
-   * \param sorted_list the list containing entry information
-   * \param index1,index2 the instances switched
-   * \param map_stats a vector containing the accumulated precisions for each position in a list
+   * \brief Obtain the delta MAP by trying to switch the positions of labels in ppred_idx or
+   *        npred_idx when sorted by predictions
+   * \param ppred_idx, npred_idx the position of the chosen labels when sorted by predictions
+   * \param pos_label, neg_label the chosen positive and negative labels
+   * \param p_map_stats a vector containing the accumulated precisions for each position in a list
+   * \param map_stats_size size of the accumulated precisions vector
    */
-  inline static bst_float GetLambdaMAP(const std::vector<ListEntry> &sorted_list,
-                                       int index1, int index2,
-                                       std::vector<MAPStats> *p_map_stats) {
-    std::vector<MAPStats> &map_stats = *p_map_stats;
-    if (index1 == index2 || map_stats[map_stats.size() - 1].hits == 0) {
+  XGBOOST_DEVICE inline static bst_float GetLambdaMAP(
+    int ppred_idx, int npred_idx,
```

Comment:
According to google style variable names should not be abbreviated, except for common conventions (idx, num is fine). I don't know what ppred/npred means. Longer variable names are ok.

Reply:
done